sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/join/EventTimeEquiJoin.java

```diff
@@ -0,0 +1,272 @@
+        Coder<V1> firstValueCoder,
+        Coder<V2> secondValueCoder,
+        Duration firstValidFor,
+        Duration secondValidFor,
+        Duration allowedLateness) {
+      this.firstCollectionValidFor = firstValidFor;
+      this.secondCollectionValidFor = secondValidFor;
+      this.allowedLateness = allowedLateness;
+      firstCollectionItems = StateSpecs.orderedList(firstValueCoder);
+      secondCollectionItems = StateSpecs.orderedList(secondValueCoder);
+    }
+
+    @FunctionalInterface
+    private interface Output<T1, T2> {
+      void apply(T1 one, T2 two);
+    }
+
+    private <T, O> void processHelper(
+        Output<T, O> output,
+        KV<K, RawUnionValue> element,
+        Instant ts,
+        OrderedListState<T> thisCollection,
+        OrderedListState<O> otherCollection,
+        TimerMap cleanupTimers,
+        Duration thisCollectionValidFor,
+        Duration otherCollectionValidFor) {
+      thisCollection.add(TimestampedValue.of((T) element.getValue().getValue(), ts));
+      Instant beginning = ts.minus(otherCollectionValidFor);
+      Instant end = ts.plus(thisCollectionValidFor).plus(1L);
+      for (TimestampedValue<O> value : otherCollection.readRange(beginning, end)) {
+        output.apply((T) element.getValue().getValue(), value.getValue());
+      }
+      Instant cleanupTime = ts.plus(allowedLateness).plus(thisCollectionValidFor);
+      Instant nextBucketStart =
+          Instant.ofEpochMilli(
+              cleanupTime.getMillis() / TIMER_BUCKET * TIMER_BUCKET + TIMER_BUCKET);
+      cleanupTimers.get(Long.toString(nextBucketStart.getMillis())).set(nextBucketStart);
+    }
+
+    @ProcessElement
+    public void process(
+        ProcessContext context,
+        @Element KV<K, RawUnionValue> element,
+        @Timestamp Instant ts,
+        @StateId("v1Items") OrderedListState<V1> firstItems,
+        @StateId("v2Items") OrderedListState<V2> secondItems,
+        @TimerFamily("cleanupTimers") TimerMap cleanupTimers) {
+      switch (element.getValue().getUnionTag()) {
+        case FIRST_TAG:
+          processHelper(
+              (V1 v1, V2 v2) -> {
+                context.output(KV.of(element.getKey(), Pair.of(v1, v2)));
+              },
+              element,
+              ts,
+              firstItems,
+              secondItems,
+              cleanupTimers,
+              firstCollectionValidFor,
+              secondCollectionValidFor);
+          break;
+        case SECOND_TAG:
+          processHelper(
+              (V2 v2, V1 v1) -> {
+                context.output(KV.of(element.getKey(), Pair.of(v1, v2)));
+              },
+              element,
+              ts,
+              secondItems,
+              firstItems,
+              cleanupTimers,
+              secondCollectionValidFor,
+              firstCollectionValidFor);
+      }
+    }
+
+    @OnTimerFamily("cleanupTimers")
+    public void onCleanupTimer(
+        @TimerId String timerId,
```

Comment:
You could add a @Timestamp parameter, you don't need to parse it out of the string timerId